core/shared/src/main/scala/zio/stm/TMap.scala

```diff
@@ -0,0 +1,223 @@
+    }
+
+    for {
+      buckets <- tBuckets.get
+      idx     <- indexOf(k)
+      _       <- buckets.updateM(idx, removeMatching)
+    } yield ()
+  }
+
+  /**
+   * Atomically folds using pure function.
+   */
+  final def fold[A](zero: A)(op: (A, (K, V)) => A): STM[Nothing, A] =
+    for {
+      buckets <- tBuckets.get
+      res     <- buckets.fold(zero)((acc, bucket) => bucket.foldLeft(acc)(op))
+    } yield res
+
+  /**
+   * Atomically folds using effectful function.
+   */
+  final def foldM[A, E](zero: A)(op: (A, (K, V)) => STM[E, A]): STM[E, A] = {
+    def loopM(acc: STM[E, A], remaining: List[(K, V)]): STM[E, A] =
+      remaining match {
+        case Nil          => acc
+        case head :: tail => loopM(acc.flatMap(op(_, head)), tail)
+      }
+
+    for {
+      buckets <- tBuckets.get
+      res     <- buckets.foldM(zero)((acc, bucket) => loopM(STM.succeed(acc), bucket))
+    } yield res
+  }
+
+  /**
+   * Atomically performs side-effect for each binding present in map.
+   */
+  final def foreach[E](f: ((K, V)) => STM[E, Unit]): STM[E, Unit] =
+    self.foldM(())((_, kv) => f(kv))
+
+  /**
+   * Retrieves value associated with given key.
+   */
+  final def get(k: K): STM[Nothing, Option[V]] =
+    for {
+      buckets <- tBuckets.get
+      idx     <- indexOf(k)
+      bucket  <- buckets(idx)
+    } yield bucket.find(_._1 == k).map(_._2)
+
+  /**
+   * Retrieves value associated with given key or default value, in case the
+   * key isn't present.
+   */
+  final def getOrElse(k: K, default: => V): STM[Nothing, V] =
+    get(k).map(_.getOrElse(default))
+
+  /**
+   * Stores new binding into the map.
+   */
+  final def put(k: K, v: V): STM[Nothing, Unit] = {
+    def upsert(bucket: List[(K, V)]): STM[Nothing, List[(K, V)]] = {
+      val exists = bucket.find(_._1 == k).isDefined
+
+      if (exists)
+        STM.succeed(bucket.map(kv => if (kv._1 == k) (k, v) else kv))
+      else
+        tSize.update(_ + 1).as((k, v) :: bucket)
+    }
+
+    def resize(newCapacity: Int): STM[Nothing, Unit] =
+      for {
+        data       <- self.fold(List.empty[(K, V)])((acc, kv) => kv :: acc)
+        tmap       <- TMap.allocate(newCapacity, data)
+        newBuckets <- tmap.tBuckets.get
+        _          <- self.tBuckets.set(newBuckets)
+        _          <- self.tCapacity.set(newCapacity)
+      } yield ()
+
+    for {
+      buckets     <- tBuckets.get
+      idx         <- indexOf(k)
+      _           <- buckets.updateM(idx, upsert)
+      size        <- tSize.get
+      capacity    <- tCapacity.get
+      needsResize = capacity * TMap.LoadFactor < size
+      _           <- if (needsResize) resize(capacity * 2) else STM.unit
+    } yield ()
+  }
+
+  /**
+   * Atomically updates all bindings using pure function.
+   */
+  final def transform(f: ((K, V)) => (K, V)): STM[Nothing, Unit] =
+    for {
+      data       <- self.fold(List.empty[(K, V)])((acc, kv) => f(kv) :: acc)
```

Comment:
This code is probably duplicated a few times and should be a `toList` or something method. Also we don't have to fix it here, of course, but in theory, you can be _much_ more efficient, because you can reuse the same `TArray` and just change the entries.

Reply:
Something like this?

```scala
  final def transform(f: (K, V) => (K, V)): STM[Nothing, Unit] =
    for {
      data     <- self.fold(List.empty[(K, V)])((acc, kv) => f(kv._1, kv._2) :: acc)
      buckets  <- self.tBuckets.get
      capacity <- self.tCapacity.get
      _        <- buckets.transform(_ => Nil)
      updates  = data.map(kv => buckets.update(kv._1.hashCode() % capacity, kv :: _))
      _        <- STM.collectAll(updates)
    } yield ()
```

Reply:
Yes, exactly.